[Libcloud-589] ProfitBricks Compute Driver Support

baldwinSPC opened the pull request:
I have completed support for the ProfitBricks compute driver. This adds support for the core compute driver functions against ProfitBricks. The JIRA ticket can be found here:

https://issues.apache.org/jira/browse/LIBCLOUD-589

Let me know if there are any issues or if it can be accepted or not.

Thanks. 
-matt

baldwinSPC committed:
Pushing the PB compute driver into branch for PR to the Apache Libcloud project.

baldwinSPC committed:
[Libcloud-589] ProfitBricks Compute Driver Support

Kami commented:
Sweet, thanks!

I will try to review it asap.

Kami commented:
Travis build appears to be failing - https://travis-ci.org/apache/libcloud/builds/33635636

baldwinSPC commented:
Thanks. Going through and correcting them now.

On Tue, Aug 26, 2014 at 2:28 PM, Tomaz Muraus notifications@github.com
wrote:

> Travis build appears to be failing -
> https://travis-ci.org/apache/libcloud/builds/33635636
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/apache/libcloud/pull/352#issuecomment-53495508.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1544 @@
+            datacenter) for datacenter in object.findall('.//return')]
+
+    def _to_datacenter(self, datacenter):
+        elements = list(datacenter.iter())
```
`Element.iter` is not available in Python 2.6 (https://docs.python.org/2/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element.iter), so please use `findall` instead (`datacenter.findall('.//*')`).

baldwinSPC replied:
@Kami OK, this is corrected now: all spots where I use iter are now using findall.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1544 @@
+from libcloud.compute.types import NodeState
+from libcloud.common.types import LibcloudError, MalformedResponseError
+
+__version__ = '1.0.0'
```
We don't `__version__` tag / attribute in driver modules, so please remove it.

baldwinSPC replied:
@Kami I've removed the version line.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1544 @@
+import copy
+import time
+
+# from xml.etree import ElementTree as ET
```
Seems like this line can be removed all together?

baldwinSPC replied:
@Kami Done. The line has been removed.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1544 @@
+        try:
+            body = ET.XML(self.body)
+        except:
+            raise MalformedResponseError("Failed to parse XML",
```
Minor style thing - for consistency, please use single quotes around strings.

baldwinSPC replied:
@Kami Strings are now single quoted.

baldwinSPC replied:
@Kami The only issue I have at the moment is with Travis CI. It seems that I'm failing the py3x tests, but all others pass. I'm not sure how to diagnose the issue as it refers to the test_profitbricks module. 

AttributeError: 'module' object has no attribute 'test_profitbricks'
ERROR: InvocationError: '/home/travis/build/apache/libcloud/.tox/py32/bin/python setup.py test'

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

baldwinSPC committed:
[Libcloud-589] Correcting issues raised in Travis CI.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+
+        if 'request' in data.keys():
+            soap_req_body = ET.SubElement(soap_req_body, 'request')
+            for key, value in data.iteritems():
```
`.iteritems()` is not available in Python 3, so you should just use `.items()`.

baldwinSPC replied:
Code uses .items() now.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+        if 'request' in data.keys():
+            soap_req_body = ET.SubElement(soap_req_body, 'request')
+            for key, value in data.iteritems():
+                if not (key == 'action' or key == 'request'):
```
It would be a bit more "pythonic" to write `if key not in ['action', 'request']`.

baldwinSPC replied:
This is corrected.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+                    child.text = value
+        else:
+            for key, value in data.iteritems():
+                if not (key == 'action'):
```
`if key != 'action'`

baldwinSPC replied:
done.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+                                                           raw=raw)
+
+
+class Datacenter(UuidMixin):
```
Hm, it seems like you could use existing `libcloud.compute.baseNodeLocation` base class instead of defining a new Profitbricks specific class?

Or is this not possible?

baldwinSPC replied:
@Kami I have fixed most of what you have raised. I'm now beginning to look at the Datacenter class to see if I can re-use the NodeLocation class. At the moment, I'm not sure if this would be the right approach. The ProfitBricks platform requires you to create a virtual datacenter (VDC) within which you place your instances. This is similar to the Cloud Service concept on Azure. The docs for NodeLocation indicate that this is the physical location of the nodes. A VDC is a virtual location. 

I agree that my Datacenter class looks basically like the NodeLocation class. My concern is with usability and users of ProfitBricks understanding that location equates to a virtual datacenter. Additionally, the VDC has a region versus country associated to it, e.g. Las Vegas vs United States. NodeLocation uses 'country'. 

As I expand this driver I will expand the Datacenter class to incorporate more features they support on the VDCs, e.g. doing a GET on a VDC will show you all objects attached to that VDC. This seems to all go beyond the scope of NodeLocation. 

Thoughts?

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+        sizes = []
+
+        for key, values in self.PROFIT_BRICKS_GENERIC_SIZES.items():
+            node_size = self._to_node_size(copy.deepcopy(values))
```
It looks like `_to_node_size` method doesn't manipulate an input dictionary so `copy.deepcopy` is redundant.

baldwinSPC replied:
done.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+        locations = []
+
+        for key, values in self.REGIONS.items():
+            location = self._to_location(copy.deepcopy(values))
```
Same here (`copy.deepcopy` is redundant).

baldwinSPC replied:
Cleaned up.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+
+    def create_node(self, name, image, size=None, volume=None,
+                    datacenter=None, internet_access=True,
+                    availability_zone=None, **kwargs):
```
Instead of using `**kwargs`, please explicitly declare all the available arguments[[1](https://libcloud.readthedocs.org/en/latest/development.html#don-t-abuse-kwargs)] and prefix them the ones which are not part of the base API with `ex_`.

E.g.

``` python
def create_node(self, name, image, size=None, location=None, ex_availability_zone=None, ex_internet_access=True, ex_ram=None, ex_cores=None, ex_disk=None)
```

As you can see, I also used `location` instead of `datacenter` (this refers to the comment about NodeLocation I've mentioned above).

baldwinSPC replied:
made the change.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+                'location': location.lower()
+                }
+
+        if not name:
```
This check is not really necessary since `name` argument in the method signature doesn't have a default value.

baldwinSPC replied:
Removed.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+                                    data=body,
+                                    method='POST').object)
+
+    def ex_update_datacenter(self, datacenter, name):
```
It seems like this just renames a datacenter? If so, it should probably just be called `ex_rename_datacenter`.

baldwinSPC replied:
Updated the function.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+        """
+        Update a datacenter.
+
+        :param datacenter: The DC you're destroying.
```
Need to update the docstring.

baldwinSPC replied:
Updated the docstring.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+
+        This removes all objects in a DC.
+
+        :param datacenter: The DC you're destroying.
```
Same here.

baldwinSPC replied:
Ditto

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+                'nicId': network_interface.id
+                }
+
+        return self._to_interfaces(
```
It seems like this function will return a list, but the method name and docstring say it returns a single objects.

Seems like it should call `self._to_interface` or do `interface = self._to_interfaces(...)[0]` instead? It would also be good to add a test case for it.

baldwinSPC replied:
Changed this to return a single object correctly. There's also a test case for ex_describe_network_interface.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+
+        return True
+
+    """ Snapshot Functions Not Implemented
```
This can be removed since `create_volume_snapshot` in the base class already throws `NotImplementedError.`

baldwinSPC replied:
Removed.

Kami commented on libcloud/test/compute/test_profitbricks.py:
```diff
@@ -0,0 +1,520 @@
+# limitations under the License.
+
+import sys
+import httplib
```
This should be `from libcloud.utils.py3 import httplib`.

This import is also the reason why tests are failing under Python 3.

baldwinSPC replied:
Updated.

Kami commented on libcloud/test/compute/test_profitbricks.py:
```diff
@@ -0,0 +1,520 @@
+
+    libcloud.security.VERIFY_SSL_CERT = False
+
+    USER_ID = 'user@profitbricks.com'
```
For consistency, please use credentials from `secrets.py-dist` file. You can check existing test files for an example of how to do that.

baldwinSPC replied:
Updated.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+from libcloud.compute.base import UuidMixin
+from libcloud.compute.types import NodeState
+from libcloud.common.types import LibcloudError, MalformedResponseError
+
```
Please declare all the "exported" / "public" classes using `__all__`.

E.g.

``` python
__all__ = [
    'ProfitBricksNetworkInterface`,
    'ProfitBricksNodeDriver`,
   ....
]
```

baldwinSPC replied:
Added **all** in.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+    Base ProfitBricks node driver.
+    """
+    connectionCls = ProfitBricksConnection
+    name = "ProfitBricks Node Provider"
```
Minor style thing - single quotes

baldwinSPC replied:
Corrected.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+        :rtype:     :class:`Node`
+        """
+
+        if not datacenter:
```
To make this method more readable and easier to maintain, it would be good to split it into multiple smaller "private" methods.

For example, this part could be split into `def _create_new_datacenter_for_node`, `def _wait_for_datacenter_state(datacenter, state=PROVISIONING_STATE.AVAILABLE, timeout=300, interval=5)`, etc.

baldwinSPC replied:
This is refactored.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1531 @@
+        else:
+            password = None
+
+        '''
```
Same here, this part could be refactored into `_create_volume` and `_wait_for_volume_state` ...

baldwinSPC replied:
Same here. This is refactored, too.

Kami commented on libcloud/test/compute/test_profitbricks.py:
```diff
@@ -0,0 +1,520 @@
+                                 dict(
+                                 id="cd59b162-0289-11e4-9f63-52540066fee9"))
+
+        describe = self.driver.ex_describe_network_interface(network_interface=network_interface)[0]
```
As mentioned above, this method shouldn't return a list :)

baldwinSPC replied:
Corrected.

Kami commented:
Alright, done with a more thorough review.

Overall it looks pretty solid, it just need a bit more cleanup and refactoring work :)

Kami commented on libcloud/test/compute/test_profitbricks.py:
```diff
@@ -0,0 +1,520 @@
+
+class ProfitBricksTests(unittest.TestCase):
+
+    libcloud.security.VERIFY_SSL_CERT = False
```
This should be unnecessary for the tests.

baldwinSPC replied:
Removed this, too.

baldwinSPC committed:
squashing all feedback items.

baldwinSPC committed:
[Libcloud-589] correcting linting and py3 issue.

baldwinSPC committed:
[Libcloud-589] correcting linting and py3 issue.

baldwinSPC commented:
@Kami I've corrected all items you called out in your review. The only pending question is the Datacenter class. Let me know what you're thoughts are now. Travis CI passes all tests. :+1:

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+    """
+    def __init__(self, id, name, datacenter_version, driver, extra=None):
+        self.id = str(id)
+        if name is None:
```
Minor thing - this if is redundant. Just doing `self.name = name` is enough.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+        self.name = name
+
+    def __repr__(self):
+        return (('<ExProfitBricksAvailabilityZone: name=%s>')
```
Should remove `Ex` from **repr** string since class doesn't have this prefix.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+        self.extra = extra or {}
+
+    def __repr__(self):
+        return (('<ProfitBricksNetworkInterface: id=%s, name=%s')
```
Missing closing `>` in the string.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+
+    Note: This class is ProfitBricks specific.
+    """
+    def __init__(self, id, name, datacenter_version, driver, extra=None):
```
This class represents a datacenter and other arguments don't have a `datacenter_` prefix, so this one should just be called `version` to be consistent with other arguments.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+        return self._to_nodes(self.connection.request(action=action,
+                              data=body, method='POST').object)
+
+    def reboot_node(self, node=None):
```
`node` argument is mandatory so it shouldn't default to `None` (this method also doesn't make sense if the `node` argument is not provided).

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+                                                      data=body,
+                                                      method='POST').object)
+
+    def destroy_node(self, node, remove_attached_disks=None):
```
Should default to `False` and not `None`.

Kami replied:
Also, it's an extension argument so it should be prefixed with `ex_`.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+                                                        data=body,
+                                                        method='POST').object)
+
+    def attach_volume(self, node, volume, device=None, bus_type=None):
```
`bus_type` is an extension argument so it should be prefixed with `ex_`.

Kami replied:
Also, what happens if `None` is sent as the value for the `busType` attribute? Does it use a server side default value or does it error out?

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+                                                        data=body,
+                                                        method='POST').object)
+
+    def detach_volume(self, node, volume):
```
Base API method doesn't include `node` argument.

Is it possible to retrieve the id of the node the volume is attached to by retrieving information for a single volume from the API? 

This way we can follow the standard API (`def detach_volume(self, volume)`) and don't need to change the method signature (`def detach_volume(self, volume, ex_node)`).

Kami replied:
Edit: User is already expected to pass in an instance of the `Volume` class so we don't actually need to hit the API.

Just doing `node_id = volume.extra['server_id']` and then using that in the request body should work.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+        """
+        Detaches a volume.
+
+        :param volume: The volume you're attaching.
```
Docstrings need to be updated (volume to be detached).

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+
+        return True
+
+    def ex_describe_volume(self, volume):
```
The method should take `volume_id` and not `volume` (and instance of the `Volume` class) argument.

The user should already have all the volume information if it has a reference to the `volume` object. Unless of course, the Volume class is instantiated manually with just an id, but that's an invalid use of the API in this case.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+        """
+        Describes a volume.
+
+        :param volume: The volume you're attaching..
```
Docstring needs to be updated.

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+        return True
+
+    def ex_set_inet_access(self, datacenter,
+                           network_interface, ex_internet_access=True):
```
This is an extension method, so the argument itself doesn't need to have `ex_` prefix (just `internet_access` is fine).

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+                else:
+                    public_ips.append(ip)
+
+        if ET.iselement(node.find('cpuHotPlug')):
```
It seems like we could reduce a couple of lines of code and make this a bit easier to maintain by using a dict and a for loop.

Something along the lines of:

``` python
attribute_to_extra_name_map = {
    'ramHotPlug': 'memory_hot_pluggable',
    ...
}

for attribute_name, extra_name in attribute_to_extra_name_map.items():
    if ET.iselement(node.find(attribute_name)):
        value = node.find(attribute_name).text
    else:
        value = None

    extra[extra_name] = value
```

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+        """
+        dc_operation_status = self.ex_describe_datacenter(datacenter[0])
+
+        timeout = 60 * 5
```
Minor thing - it would be nice if `timeout`, `wait_time` and `interval` variables would be declared in the method signature with the same default values which are currently used.

Same goes for `_wait_for_storage_volume_state`

Kami commented on libcloud/compute/drivers/profitbricks.py:
```diff
@@ -0,0 +1,1569 @@
+
+            waittime += interval
+            time.sleep(interval)
+
```
This method doesn't actually do anything different if a timeout occurs (`None` is implicitly returned in both cases).

On success, we should probably return object in question (latest reference to the `datacenter` object in this case) and on time out, we should throw a timeout exception.

Same goes for `_wait_for_storage_volume_state`.

Kami replied:
While reviewing the changes, I also just noticed `ex_describe_datacenters` incorrectly returns a list. Anyway, I'll fix it.

Kami commented:
Added some more comments.

When those are addressed, please squash all the commits.

Kami commented:
@baldwinSPC How are things looking for those last couple of issues?

I will probably be preparing a new release this weekend and it would be great if those things could get addressed so I can also include this driver.

If you can't do it this weekend, let me know and I will do it myself (address those issues + squash the commits) while merging the patch.

baldwinSPC commented:
Hey Tomaz

I've corrected about half of these. I'll commit and push those this
afternoon. I don't think I can correct the others, but what I'll do is
comment on the ones I've finished up.
On Sep 6, 2014 6:30 AM, "Tomaz Muraus" notifications@github.com wrote:

> @baldwinSPC https://github.com/baldwinSPC How are things looking for
> those last couple of issues?
> 
> I will probably be preparing a new release this weekend and it would be
> great if those things could get addressed so I can also include this driver.
> 
> If you can't do it this weekend, let me know and I will do it myself
> (address those issues + squash the commits) while merging the patch.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/apache/libcloud/pull/352#issuecomment-54712319.

baldwinSPC committed:
corrects most of the feedback, but not quite all.

baldwinSPC committed:
changed ex_describe_volume to take volume_id versus a volume.

baldwinSPC committed:
correcting lint issue.

baldwinSPC committed:
Removed 'node' from the param list of detach_volume.

baldwinSPC commented:
@Kami I just pushed some updates that address most of the items you called out -- there are only two remaining ones: waiting for state and re-working the ET code.

Kami commented:
@baldwinmathew Great, thanks.

I'll work on rest of the issues and try to get it merged tonight.

Kami commented:
Another thing which got me confused while reviewing the changes:

``` python
PROVISIONING_STATE = {
        'INACTIVE': NodeState.PENDING,
        'INPROCESS': NodeState.PENDING,
        'AVAILABLE': NodeState.RUNNING,
        'DELETED': NodeState.TERMINATED,
    }

....

def _wait_for_datacenter_state(
            self,
            datacenter,
            state=PROVISIONING_STATE.get(NodeState.RUNNING),
            timeout=300,
            interval=5):
```

`NodeState.RUNNING` is mapped to int `0` (`libcloud.compute.types.NodeState`), but `PROVISIONING_STATE` dictionary contains strings as keys. Unless I am missing something, this means `.get` will always return `None` and `_wait` function will always time out.

Same goes with while condition - `self.PROVISIONING_STATE.get(NodeState.PENDING)`.

I believe the `state` should default to `state=PROVISIONING_STATE['available']`. I will push those fixes to my branch and wait for you to review them before merging them into trunk.

Kami committed:
Various fixes in the Profitbricks driver:

1. Fix "ex_describe_datacenter" to return a single object instead of a list
2. Fix "ex_describe_datacenter" method signature to take an id instead of an object
3. Fix default value for "state" argument in the "_wait_for_datacenter_state" method
4. Fix while condition in the "_wait_for_datacenter_state" method
5. Modify "_wait_for_datacenter_state" to throw an Exception on timeout
6. Do the same fixes in the "_wait_for_storage_volume_state" method
7. Fix "ex_create_datacenter" to return single object instead of a list
8. Fix "create_volume" to return a single object instead of a list
9. Fix "ex_describe_node" to return a single object instead of a list
10. Fix "create_node" to retirn a single object instead of a list
11. Fix "attach_volume" to return volume on success
12. Fix "ex_create_network_interface" to return a single object instead of a list

Kami commented:
@baldwinSPC I performed another review. During the review I still spotted many issues (mostly with function claiming it returns a single object but it actually returned a list) which I have fixed in my branch - https://github.com/Kami/libcloud/commit/28e9774b601c4081d86c9ac4c6b3dd619d1470ea

I would appreciate if you can go over my fixes, double check them and test them with the actual Profitbricks API (I don't have an account).

Thanks!

baldwinSPC commented:
@Kami Great. Thanks Tomaz. I will be testing through this tomorrow morning and will update this thread.

Commit:
Merge pull request #1 from Kami/LIBCLOUD-589_profitbricks_compute_driver

Various fixes in the Profitbricks driver:

baldwinSPC commented:
@Kami I just merged your branch into mine after testing all the methods live against the API. Everything looks good, however, in my testing I caught two bugs due to empty data coming back from their API. Those are two minor changes you should see on my next commit.

baldwinSPC commented:
@Kami BTW, thank you for all your support. :)

baldwinSPC committed:
fixed two minor issues when the pb api returns no data.

Kami commented:
@baldwinmathew Great and you are welcome.

I will look into your changes and and look into merging this pull request into trunk this weekend :)

Kami commented:
Looks good.

I've made "a bunch of ifs" to "for loop" change and merged patch into trunk. Thanks!
